### PR TITLE
Friendly-api: organizations

### PR DESCRIPTION
**Pull request by maiste**
This PR provides a `Friendly_api` module which contains a friendly way to interact with the Equinix API.

It is a part of #53.

### CODE REVIEW

**Comment by maiste** [comment]
:warning:  This PR must wait for #67 to be merged before being merged!

**Comment by art-w** [comment]
Do we need both the "json api" and the "friendly api"? I'm not sure we need the raw json if it has been parsed into an ocaml record :)

File: src/lib/equinoxe/equinoxe.ml
Comment (art-w) [inline src/lib/equinoxe/equinoxe.ml:334]: I'm not a big fan of helper functions at this point:
- it gives the illusion that this is an efficient operation
- reading code is harder as this `get` is less standard than a `List.find_opt` (what error will `get` raise if it doesn't find anything?)
- it doesn't really shorten the code compared to the expected:

```ocaml
let* orga = List.find_opt p =|< get_all t in
match orga with Some found -> ... | None -> ...
```

File: src/lib/equinoxe/equinoxe_intf.ml
Comment (art-w) [inline src/lib/equinoxe/equinoxe_intf.ml:192]: There are a bunch of different ids in the Equinix API (orga, projects, devices, ...). This is not very important for organization, more a note for later: Perhaps we could keep the different `id` types private and separate to clarify the API?

```ocaml
type id
type config = { id : id ; ... }
val delete : Orga.id -> unit io
```

File: src/lib/equinoxe/equinoxe.ml
Comment (art-w) [inline src/lib/equinoxe/equinoxe.ml:326]: It would probably be easier to use the Ezjsonm lib directly than to redefine the functions `extract_xxx_from`:

```ocaml
let config_of_json j : config =
  { name = Ezjsonm.(find j ["name"] |> get_string) ; ... }

let get_all t =
  let* json = http_blabla t in
  try return @@ Ezjsonm.get_list orga_of_json json
  with Ezjsonm.Parse_error (value, err) ->
    fail (Printf.sprintf "Orga.get_all: parse error %S on %s from %s" err (to_string value) (to_string json))
```
The bonus is that the user gets the name of function that failed. And in fact, this parse error is more indicative of a bug with our library than something the user is expected to fix (so perhaps we should raise something else than \``Msg`?).

I'm not against introducing helper functions where Ezjsonm doesn't help. For example, the `Ezjsonm.find` function raises `Not_found` which doesn't fit with my proposed try/catch:

```ocaml
let get field t =
  try Ezjsonm.find t [field]
  with Not_found ->
    raise Ezjsonm.Parse_error (t, Printf.sprintf "get %S" field)
```

File: src/lib/equinoxe/equinoxe_intf.ml
Comment (maiste) [reply src/lib/equinoxe/equinoxe_intf.ml:192]: Indeed, that's a good idea as it will avoid giving the wrong `id` when we will need it.

File: src/lib/equinoxe/equinoxe.ml
Comment (maiste) [reply src/lib/equinoxe/equinoxe.ml:334]: The goal was to provide an equivalent for the  `get_organization_id` as this one directly triggers the *API*. 
Considering the introduction of the `id` type, I think I can transform it to:
```ocaml
val get : Orga.id -> Orga.confif
```
What do you think?

**Comment by maiste** [comment]
> Do we need both the "json api" and the "friendly api"? I'm not sure we need the raw json if it has been parsed into an OCaml record :)

As there is a lot of information in the JSON we are not going to provide through the `Friendly_api` types, I think it is important if people want to do more stuff :wink:

File: src/lib/equinoxe/equinoxe.ml
Comment (maiste) [reply src/lib/equinoxe/equinoxe.ml:326]: You are right, it will simplify the way to track errors if we have to fix some bugs. I need to rely more on `Ezjsonm`, instead of reinventing the wheel! :sweat_smile:

Comment (maiste) [reply src/lib/equinoxe/equinoxe.ml:326]: For the error we could extend the the type in backend with an `Internal_error of string. Thanks to it, the user would know that it is related to an internal issue with the library.

File: src/lib/equinoxe/equinoxe.ml
Comment (art-w) [reply src/lib/equinoxe/equinoxe.ml:334]: Yes if it maps directly to an Equinix endpoint! We'll need to provide some `id_of_string`/`string_of_id` to make it usable :)

File: src/lib/equinoxe/equinoxe.ml
Comment (art-w) [reply src/lib/equinoxe/equinoxe.ml:326]: Then the error could be a bit more structured than a string! (so that we don't have to inline "print" the json in the message)

**Comment by art-w** [comment]
Then that's an opportunity for a PR if something is missing in the Equinoxe binding ^^

**Comment by maiste** [comment]
It is true! But I still don't see the issue with providing the `Raw` Api? If it the fact that it seems to be the main thing to use, we can just hide it under a private module?

**Comment by art-w** [comment]
Yes a `Raw` api is fine! To clarify, it would be nice if the `Equinoxe_cohttp` module could be used directly, rather than having a nested `Equinoxe_cohttp.Api` or the incoming `Equinoxe_cohttp.Friendly_api` :)

**Comment by maiste** [comment]
Oh ok! It was a quiproquo: no problem with that :smile: 

I was wondering if instead of taking a backend, the `API` (named currently `Friendly_api` couldn't take a `Raw` instead of a Backend? 

It would be a strict application of the logic suggested above. The structure would be:
```ocaml
module Raw(B : Backend) : sig
  module Backend = B (* Required for the io *)
 (* ... *)
end

module Api(R : Raw) : sig
 (* ... *)
end
```

**Comment by art-w** [comment]
Uh well that only makes sense when you have different implementations of `Raw`. If you are always going to call `Api(Raw(...))` and never `Api(Something_else)`, then I think it's easier to understand the code when it is flatter with less indirections:

```ocaml
module Make (B : Backend) = struct
  module Raw = struct
    let blah : ... -> json io = ...
  end

  let blah : ... -> truc io = ... Raw.blah ...
end
```

(I really think we could skip the `Raw` module for extra-clarity, it's not like its functions have the "raw" form `json -> json io` even though we might not handle all the possible json inputs)

**Comment by maiste** [comment]
I understand your point of view :sweat_smile:  I feel like if we drop the `Raw` module we will no longer have a clear separation between the functions. The same function will gather the json, parse the json, put it into a specific type.
Whereas, with a module only in charge of gathering the json and another in charge of treating them, each one have a specific "function"... I don't know if you see what I have a problem with? :/

If it seems to be a good trade off for your, I'm completely ok with putting the Raw module inside the `Make` functor :smile: 
I'm even ok to make it disappear from the API (even it cost nothing to keep it).

**Comment by art-w** [comment]
It's not that important :P I feel the `Raw` module is creating more work than necessary by duplicating/splitting everything. Until Equinix starts documenting their json output, I doubt anyone is going to try the `Raw` api to check in the json that we didn't miss some stuff in the `Friendly_api`.

When you call into C bindings, you don't really expect a `void*` or `'a` as output... the raw `json` is exactly that, it could be anything... so it's our job to make it a first class caml value, to save time for others. I don't think that the `Raw` module splits the logic into a more understandable outcome, it's a bit arbitrary to only have the encoding of the input.

```
Equinoxe                    
ocaml value ---------> json ------http----> Equinix
    |                                     |
    v                                     v
ocaml value <--------- json <-----http-----
```
(... everything looks like a categorical argument when you draw arrows, well... if you could typecheck json and guarantee the encoding/decoding of any caml value, then we would have `val to_json : 'a -> 'a json` and `val of_json : 'a json -> 'a`, then `map : ('a json -> 'b json) -> 'a -> 'b = fun f x -> of_json (f (to_json x))` and tada we have a functor from Caml to REST... so I would rather split the http requests and keep the parse/print together!)

But anyway, it really doesn't matter. If you feel strongly about the `Raw` module then that's fine! :D

**Comment by maiste** [comment]
I didn't think of it in term of `void` :smile: And it totally makes sens!

This is a pretty convincing argument you are giving me here (as you said, everything feel more categorical).  Thanks to this discussion, you have convinced me to remove the `Raw` module. It will be for the `0.2.0` version of `equinoxe`. However, it would be nice if we only remove it once everything has been moved to `Friendly API`!

**Comment by maiste** [comment]
@art-w, I push a new version that takes care of all of your remarks :+1: 
It can be merged as soon as it's ok for you!

File: src/lib/equinoxe/equinoxe_intf.ml
Comment (art-w) [inline src/lib/equinoxe/equinoxe_intf.ml:181]: small typo ^^

Comment (maiste) [reply src/lib/equinoxe/equinoxe_intf.ml]: Corrected :)

**Review by art-w — APPROVED** [review]

**Pull request merged**
